[LM-1357] Initial implementation 0.1.0

dmitry-vychikov opened the pull request:
## PR Checklist
 * Which ticket is this associated with?
    * Closes LM-1357
 * Where can we test this?
    * http://SUBDOMAIN.dev.rentdynamics.com
 * Additional Notes
    * NOTES


## Dev Checklist (Code Reviewer is responsible for filling this out or providing a brief explanation why it isn't necessary)

- [x] Verify that code is tested appropriately 
- [x] Verify that all critical algorithms work as intended

Frontend Concerns
- [ ] Evaluate all calls to the DB are efficient
- [ ] Ensure nullable values are treated as such and handled appropriately
- [ ] Ensure any values returned for the DB are cached and reused
- [ ] Evaluate any code that is run by a Pusher event is very clean and very necessary



## Q/A Checklist (QA is responsible for filling this out or providing a brief explanation why it isn't necessary)
- [ ] Appropriate test plan has been updated with all testing performed
- [ ] Review that solution matches JTBD

Frontend Concerns
- [ ] Review that solution matches mockup
- [ ] Verify solution doesn't allow rage clicking of buttons

davegines commented on README.md:
```diff
@@ -0,0 +1,275 @@
+# RentDynamics.RdClient
```
Why are there two README.md files? One is all uppercase, the other is all lowercase.

dmitry-vychikov replied:
It happened when I tried to create an empty `master` branch. At that time, I created `readme.md` file just to initialize the repository with something. However, in `dev` branch we already had `README.md` file.

I remember resolving git conflicts when merging these 2 files together and it is a surprise for me that we ended up with both files.

I have removed `readme.md` because it was older and did not contain you latest changes.

davegines reviewed (changes requested):
This is looking really good Dmitry. Great work! Just a few comments below.

Can we .gitignore the nugets folder so that we're not checking in those binaries into github

dmitry-vychikov commented on .idea/.idea.RentDynamicsCS/.idea/.gitignore:
```diff
@@ -0,0 +1,13 @@
+# Default ignored files
```
None really. I've removed that folder from the repository and added it to `.gitignore` file.

dmitry-vychikov commented:
> Can we .gitignore the nugets folder so that we're not checking in those binaries into github
That 'nugets' folder is a left-over after my testing. I have removed that folder and added it to .gitignore.

dmitry-vychikov commented on RentDynamics.RdClient.Tests/IntegrationTests/IntegrationTests.cs:
```diff
@@ -0,0 +1,205 @@
+
+            var getAddress = await apiClient.GetAsync<Dictionary<string, object>>($"addresses/{addressId}");
+
+            // getAddress.Should().
```
Thanks. Fixed.

dmitry-vychikov commented on RentDynamics.RdClient.Tests/IntegrationTests/LoginLogoutTests.cs:
```diff
@@ -0,0 +1,22 @@
+    {
+        protected override bool AutomaticAuthentication => false;
+
+        [TestMethod, Ignore]
```
No, it does not. The API request fails with 500 error code because we use by-pass user authentication credentials. Last time you mentioned that we aren't fixing the API for now, so I decided to ignore the unit test.

I've added a comment to explain why it is ignored.

As an option, we could create another RentDynamics user and put its credentials to CircleCI to use from that unit test.

dmitry-vychikov commented on RentDynamics.RdClient/Resources/LeadCards/LeadCardVM.cs:
```diff
@@ -0,0 +1,69 @@
+        {
+            if (phoneNumber == null && email == null)
+            {
+                throw new ModelValidationException($"At least {nameof(phoneNumber)} or {nameof(email)} must be specified");
```
At first, I thoght that there is no need to even send the request when we know that it is going to fail.

But avoid code duplication is also a great point. I think, you are right, and we should remove that part from the client library.

dmitry-vychikov commented on RentDynamicsCS.sln.DotSettings:
```diff
@@ -0,0 +1,11 @@
+﻿<wpf:ResourceDictionary xml:space="preserve" xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml" xmlns:s="clr-namespace:System;assembly=mscorlib" xmlns:ss="urn:shemas-jetbrains-com:settings-storage-xaml" xmlns:wpf="http://schemas.microsoft.com/winfx/2006/xaml/presentation">
```
This one is actually included on purpose. It contains some team-shared resharper settings (like `VM` abbreviation so that resharper does not suggest renaming our models to say `LeadCardVm` instead of `LeadCardVM`).

I've removed some strange lines from that file to keep only what is neccessary. Let's leave that file for now. If it gets too noisy for the developers (but I think that it should not), we can always remove it later.

dmitry-vychikov commented on RentDynamics.RdClient/Resources/LeadCards/LeadCardsResource.cs:
```diff
@@ -0,0 +1,20 @@
+namespace RentDynamics.RdClient.Resources.LeadCards
+{
+    [PublicAPI]
+    public class LeadCardsResource : BaseRentDynamicsResource
```
Sounds good. I've renamed it. Also, I will make sure the references in our other projects use the new name as well.

davegines commented:
@dmitry-vychikov 
Will you add 2 more examples of how to use the API for the `AppointmentDays` and `AppointmentTimes` endpoints in the README file? We have a client who will be using those endpoints and we want to be sure they fully understand how it is invoked and what to expect in the response.

dmitry-vychikov commented:
> @dmitry-vychikov
> Will you add 2 more examples of how to use the API for the `AppointmentDays` and `AppointmentTimes` endpoints in the README file? We have a client who will be using those endpoints and we want to be sure they fully understand how it is invoked and what to expect in the response.

1. I have implemented a new section in the readme file: https://github.com/RentDynamics/rentdynamics-cs/tree/dev#endpoint-usage-examples. 
2. I added some xml docs for `AppointmentResource` methods. Please, take a look and let me know if looks okay: https://github.com/RentDynamics/RentDynamics-cs/blob/092a8b3f7c38da77b4cb3358cae402b632b09045/RentDynamics.RdClient/Resources/Appointment/AppointmentResource.cs

dmitry-vychikov commented on README.md:
```diff
@@ -287,11 +287,11 @@ The section covers some common endpoints that are accessible with the help of `R
    int communityGroupId = <communityGroupId>;
    DateTime appointmentDate = DateTime.Today;
-   bool isUtc = true; //If set to false, the response will contain values in MDT timezone
+   bool asUtc = true; //Defaults to false. When false, times will be in local time. When true, times will be in UTC
```
> When false, times will be in local time

Is the caller going to receive the times in their local timezone? Or is it a timezone local to the API server?

davegines replied:
> > When false, times will be in local time
> 
> Is the caller going to receive the times in their local timezone? Or is it a timezone local to the API server?

When false, the times will be in the local time of the <communityGroup>. For example, if they request appointment times for a community group that is located in New York City, then it will return times in Eastern time.

dmitry-vychikov commented:
@davegines I tried to do some testing while splitting `AppointmentTimes` endpoint into 2 methods.

## `/appointmentTimes/42?appointmentDate=08/01/2020&utc=True`
Response:
```json
[
    "2020-09-16T08:00:00-0600",
    "2020-09-16T08:15:00-0600",
    "2020-09-16T08:30:00-0600",
    "2020-09-16T08:45:00-0600"
]
```

1. The response is not actually in UTC. The dates contain `-0600` offset, so it is **local** time. A date should have a zero offset to be called a UTC date. Of course, C# `DateTime` treats these values as local dates, not UTC.
2. I queried the API for `08/01/2020` appointment date, however, I got `2020-09-16` date in the response which does not seem right.

##  `/appointmentTimes/42?appointmentDate=08/01/2020&utc=False`
```json
[
    "02:00 AM",
    "02:15 AM",
    "02:30 AM",
    "02:45 AM",
]
```
The issue with that response is that we do not know which timezone it is. We imply that it is the `community local`, but which one is it actually? I'm not sure what scenarios the endpoint is used for. If our consumers don't need to know the actual timezone and what they need is just get the values and display them on the UI, then we are good. But if they need to do any datetime calculations, they might not be able to do it unless they know the timezone.


## Comparing 2 responses
2 responses should contain values corresponding to the same moments in time but represented in different formats.

Let's take the first lines of 2 responses and compare them:
`08:00:00-0600` and `02:00 AM`

Assuming that the timezone is -0600 (MDT, I believe), I need to add 6 hours to convert that date to UTC.
If I do that: 08:00 + 06:00 = 14:00 (which does not equal 02:00 AM on the other side).

In order to match 08:00 on the left to 02:00 AM on the right, we actually need to subtract 6 hours from 08:00. That would have made sense if the time on the left had `0000` offset. But it does not make any sense right now.

I think that either one of the endpoints (or even both of them) is returning not quite what we expect. Or I'm getting something wrong. Could you please shed some light on that for me @davegines ?

dmitry-vychikov commented:
@davegines  I will update the readme to contain information about update `AppointmentTimes` methods as soon as we resolve the issues I mentioned above.

davegines commented on RentDynamics.RdClient.Tests/ResourcesDataSerialization/AppointmentResourceTests.cs:
```diff
@@ -0,0 +1,78 @@
+
+            var expected = new[]
+            {
+                DateTime.Parse("2020-09-18T15:00:00+0000").ToUniversalTime(),
```
Please update format of datetime object.

dmitry-vychikov replied:
Fixed
